### ISSUE TITLE
LBFGS support

### ISSUE THREAD

**Issue by jlperla** [google-deepmind/optax#468]
Some ML models do shockingly well with classic line-search optimization methods.

Pytorch has https://pytorch.org/docs/stable/generated/torch.optim.LBFGS.html  (and pytorch lightning has a clean wrapper) which often dominates the other optimization algorithms for my problems.

The sense of a step (i.e., a `optimizer.update` and `optimizer.apply_updates` ) is a little different, of course, than in SGD/etc. but I suspect  it can be managed.  Worst case with a callback of a `jax.value_and_grad` style function.

**Comment by adam-hartshorne** [comment]
Is available in this accompanying library,

https://github.com/google/jaxopt

https://jaxopt.github.io/stable/_autosummary/jaxopt.LBFGS.html#jaxopt.LBFGS

**Comment by jlperla** [comment]
Thanks @adam-hartshorne   Is there no way that this sort of algorithm couldn't fit within the standard optax pattern in a loop (i.e. `updates, opt_state  = optimizer.update(grads, opt_state, model)` etc?

Or is the issue that there is no way to fit this into the `GradientTransformation`?

**Comment by mtthss** [comment]
It is possible in principle to fit this is a gradient transformation,
but we usually leave line search and solvers to jaxopt
and keep optax focused on stochastic optimisation for common deep learning workloads.

**Comment by carlosgmartin** [comment]
@mtthss Since jaxopt is now being merged into optax, should this issue be reopened?

**Comment by jlperla** [comment]
Thanks for bringing this up, I was going to ask again.

In general, I am finding that classic line search and 1.5th order methods can be excellent for some ML problems and this would lower the bar for experimentation.

**Comment by fabianp** [comment]
hey, there is now a pull request introducing LBFGS that will get merged soon. I encourage those interested to take a look and leave some feedback there!

**Comment by lukemarris** [comment]
Is adding LBFGS-B support (https://jaxopt.github.io/stable/_autosummary/jaxopt.LBFGSB.html) also planed?
